internal/blobtesting/map.go

```diff
@@ -98,8 +105,13 @@ func (s *mapStorage) DeleteBlob(ctx context.Context, id blob.ID) error {
 	s.mutex.Lock()
 	defer s.mutex.Unlock()
 
+	if r, ok := s.retentionTime[id]; ok && r.After(s.timeNow()) {
```

Comment:
This needs to emulate the behavior of versioned stores.
- `DeleteBlob` should add a deletion marker
- A separate `DeleteBlobVersion` should be used to delete specific version, and for that particular case then the retention period should be checked.

Reply:
I do not think there is a need yet to extend the storage api with `DeleteBlobVersion` for blob-retention testing. There are a couple other APIs (SetTime/TouchBlob) etc that will suffice for now I believe. See how I am using those in unit-tests to check for retention settings.

Ideally, you'd want to verify hard-deletes (delete-by-version) as well, but that expands the scope of this PR, so could be a whole another PR after we are done with blob-retention.

Reply:
@Shrekster The blob retention settings apply specifically to `DeleteBlobVersion` and no other blob API calls.

When `versionId` is specified in an [S3 DeleteObject call](https://docs.aws.amazon.com/AmazonS3/latest/API/API_DeleteObject.html#API_DeleteObject_RequestSyntax), then the call will fail if the retention period has not elapsed.

Reply:
@julio-lopez Correct, but I noticed we have some extended storage APIs called `SetTime` & `TouchBlob` as well to test mutations. I'd assume they would NOT behave any differently with retention when compared to `DeleteBlobVersion`. I am obviously also making an assumption that these extensions are specifically mutating the last version and NOT pushing a new version.

I am a little sensitive to adding test-only interfaces to the APIs unless needed. Let me know if you feel strongly on whether they should push a new version too, then I can go ahead and make that change.